aws-lc-rs/src/signature.rs

```diff
@@ -429,6 +572,72 @@ impl<B: AsRef<[u8]>> UnparsedPublicKey<B> {
         self.algorithm
             .verify_digest_sig(self.bytes.as_ref(), digest, signature)
     }
+
+    /// Parses the public key bytes and returns a `ParsedPublicKey`.
+    ///
+    /// This method validates the public key format and creates a `ParsedPublicKey`
+    /// that can be used for more efficient signature verification operations.
+    /// The parsing overhead is incurred once, making subsequent verifications
+    /// faster compared to using `UnparsedPublicKey::verify` directly.
+    ///
+    /// This is equivalent to calling [`ParsedPublicKey::new`] with the same
+    /// algorithm and bytes.
+    ///
+    /// # Errors
+    /// `KeyRejected` if the public key bytes are malformed or incompatible
+    /// with the specified algorithm.
+    pub fn parse(&self) -> Result<ParsedPublicKey, KeyRejected> {
+        parse_public_key(self.bytes.as_ref(), self.algorithm)
+    }
+}
+
+pub(crate) fn parse_public_key(
+    bytes: &[u8],
+    algorithm: &'static dyn VerificationAlgorithm,
+) -> Result<ParsedPublicKey, KeyRejected> {
+    let parsed_algorithm: &'static dyn ParsedVerificationAlgorithm;
+
+    let key = if algorithm.type_id() == TypeId::of::<EcdsaVerificationAlgorithm>() {
+        #[allow(clippy::cast_ptr_alignment)]
+        let ec_alg = unsafe {
+            &*(algorithm as *const dyn VerificationAlgorithm).cast::<EcdsaVerificationAlgorithm>()
+        };
```

Comment:
`(algorithm as &dyn Any).downcast_ref::<EcdsaVerificationAlgorithm>().unwrap()` seems like the logical way you are supposed to use with TypeId, and avoids the unsafe aspects.

Reply:
Yeah, that was my first approach and it worked cleanly.  The problem is that Trait upcasting (e.g., `(algorithm as &dyn Any)`) was not supported until Rust 1.86: https://blog.rust-lang.org/2025/04/03/Rust-1.86.0/#trait-upcasting

The approach I took seemed to work best, at least while we maintain 1.70 compatibility.